exec/cnex/lib/mmap_posix.c

```diff
@@ -0,0 +1,184 @@
+    Cell *r = cell_fromCString(s);
+    return r;
+}
+
+Object *object_createMMapObject(TMmapObject *p)
+{
+    Object *r = object_createObject();
+    r->ptr = p;
+    r->release = object_releaseMMapObject;
+    r->toString = object_mmapObjectToString;
+    return r;
+}
+
+
+
+
+void mmap_close(TExecutor *exec)
+{
+    Object *pf = top(exec->stack)->object;
+
+    TMmapObject *f = check_file(exec, pf);
+    if (f == NULL) {
+        pop(exec->stack);
+        return;
+    }
+
+    munmap(f->view, f->len);
+    close(f->fd);
+    f->fd = -1;
+
+    pop(exec->stack);
+}
+
+void mmap_open(TExecutor *exec)
+{
+    pop(exec->stack);
+    char *name = string_asCString(top(exec->stack)->string); pop(exec->stack);
+
+    char err[256];
+    TMmapObject *f = malloc(sizeof(TMmapObject));
+
+    f->fd = open(name, O_RDONLY);
+    if (f->fd < 0) {
+        int e = errno;
+        free(f);
+        free(name);
+        snprintf(err, 256, "open: error (%d) %s", e, strerror(e));
+        exec->rtl_raise(exec, "OpenFileException", err, BID_ZERO);
+        return;
+    }
+
+    struct stat st;
+    fstat(f->fd, &st);
+    f->len = st.st_size;
+    f->view = (char *)mmap(NULL, f->len, PROT_READ, MAP_PRIVATE, f->fd, 0);
+    if (f->view == MAP_FAILED) {
+        int e = errno;
+        close(f->fd);
+        free(f);
+        free(name);
+        snprintf(err, 256, "mmap: error (%d) %s", e, strerror(e));
+        exec->rtl_raise(exec, "OpenFileException", err, BID_ZERO);
+        return;
+    }
+
+    push(exec->stack, cell_fromObject(object_createMMapObject(f)));
+}
+
+void mmap_read(TExecutor *exec)
+{
+    Number count = top(exec->stack)->number; pop(exec->stack);
+    Number offset = top(exec->stack)->number; pop(exec->stack);
+    Object *pf = top(exec->stack)->object;  pop(exec->stack);
+    Cell *ret = cell_createStringCell(0);
+    ret->type = cBytes;
+
+    TMmapObject *f = check_file(exec, pf);
+
+    uint64_t o = number_to_uint64(offset);
+    if (o >= f->len) {
+        ret->type = cBytes;
+        push(exec->stack, ret);
+        return;
+    }
+    uint64_t c = number_to_uint64(count);
+    if (o + c > f->len) {
+        c = f->len - o;
+    }
+    ret->string = string_appendData(ret->string, (char*)f->view + o, c);
+
+    push(exec->stack, ret);
+}
+
+void mmap_size(TExecutor *exec)
+{
+    Object *pf = top(exec->stack)->object;  pop(exec->stack);
+
+    TMmapObject *f = check_file(exec, pf);
+    if (f == NULL) {
+        return;
+    }
+
+    push(exec->stack, cell_fromNumber(number_from_uint64(f->len)));
+}
+
+void mmap_write(TExecutor *exec)
+{
+    TString *data = string_fromString(top(exec->stack)->string); pop(exec->stack);
+    Number offset = top(exec->stack)->number; pop(exec->stack);
+    Object *pf = top(exec->stack)->object;  pop(exec->stack);
+
+    TMmapObject *f = check_file(exec, pf);
+    if (f == NULL) {
+        string_freeString(data);
+        return;
+    }
+    uint64_t o = number_to_uint64(offset);
+    if (o >= f->len) {
```

Comment:
I don't think you need both this and the following test. The first condition is covered by the second test, and the resulting action is the same. I think you only need the second one.

Reply:
Removed `if (o >= f->len) {` test.